### PR TITLE
[api] [fuzz] fix many header related config fuzz bugs

### PR DESCRIPTION
**Pull request by asraa**
This change includes validations on header names and values that appear in config fields. This prevents setting header keys/values with invalid characters in config fields, which pops up as ASSERT failures when converting to HeaderString values.

For reference, the well_known_regex for the header names and values do the following:
- HTTP_HEADER_NAME: whitelists alphanums and the whitelisted tokens (`!#$%&'*+-.^_|~\u0060`) in https://tools.ietf.org/html/rfc7230#section-3.2
- HTTP_HEADER_VALUE: blacklists control characters except SPC and TAB. Purposely meant to be permissive, and blacklist problems like nulls

Changes in `base.proto`
* Invalid headers to match (from [`HeaderValue`](https://github.com/asraa/envoy/blob/88d35569818cb3a1236cce5d79299c4986ffdc77/api/envoy/api/v2/core/base.proto#L234) proto)
  - `clusterfuzz-testcase-minimized-route_fuzz_test-5635252339343360`

Changes in `route_components.proto`:
* [domains](https://github.com/asraa/envoy/blob/88d35569818cb3a1236cce5d79299c4986ffdc77/api/envoy/api/v2/route/route_components.proto#L75) with invalid control characters in `VirtualHost` config
  - `clusterfuzz-testcase-minimized-route_fuzz_test-5699465522970624`
* Invalid [header_name](https://github.com/asraa/envoy/blob/88d35569818cb3a1236cce5d79299c4986ffdc77/api/envoy/api/v2/route/route_components.proto#L604) in hash policy
  - `clusterfuzz-testcase-minimized-route_fuzz_test-5634743613259776`
* Invalid control characters in [upgrade_type](https://github.com/asraa/envoy/blob/88d35569818cb3a1236cce5d79299c4986ffdc77/api/envoy/api/v2/route/route_components.proto#L698) header
  - `clusterfuzz-testcase-minimized-route_fuzz_test-5750746072481792`
* `RateLimit` message has a [header_name](https://github.com/asraa/envoy/blob/88d35569818cb3a1236cce5d79299c4986ffdc77/api/envoy/api/v2/route/route_components.proto#L1273) field
  - `clusterfuzz-testcase-minimized-route_fuzz_test-5206842068697088`
* Invalid [response_headers_to_add](https://github.com/asraa/envoy/blob/88d35569818cb3a1236cce5d79299c4986ffdc77/api/envoy/api/v2/route/route_components.proto#L113) in 
  - `clusterfuzz-testcase-minimized-route_fuzz_test-4592245302362112`
* The [cluster_header](https://github.com/asraa/envoy/blob/88d35569818cb3a1236cce5d79299c4986ffdc77/api/envoy/api/v2/route/route_components.proto#L723) change came up in a fuzz bug, but after that was fixed, the code ran in to another deeper issue to be fixed.
* The [name](https://github.com/envoyproxy/envoy/blob/bbdc33e53723dc02b6d51bb0f329c5b369adfe03/api/envoy/config/route/v3/route_components.proto#L1381) in `HeaderMatcher` message
  - `clusterfuzz-testcase-minimized-route_fuzz_test-5635252339343360`

Changes in `custom_tag.proto`
* `CustomTag` has a [name](https://github.com/asraa/envoy/blob/88d35569818cb3a1236cce5d79299c4986ffdc77/api/envoy/type/tracing/v3/custom_tag.proto#L51)
  - `clusterfuzz-testcase-minimized-route_fuzz_test-5661762636742656`

Changes in fuzz tests:
* `conn_manager_impl_fuzz_test` that fails on invalid characters in an authority header. this is not a config related change, just handled by replaced the invalid character in the fuzz test. 
   - `clusterfuzz-testcase-minimized-conn_manager_impl_fuzz_test-5714279517126656`
* `route_fuzz_test` was cleaned up now that a lot of the processing was moved to config.

Fixes:
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=19923
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=16143 (most of)

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

**Comment by lizan** [comment]
@asraa I'll review once CI is fixed.

**Comment by asraa** [comment]
Thanks! Looks like it's finally through CI. There were a few merges I had to make.

**Comment by asraa** [comment]
ping @lizan

File: api/envoy/api/v2/core/base.proto [hunk 1]
```diff
@@ -233,14 +233,17 @@ message RuntimeFeatureFlag {
 // Header name/value pair.
 message HeaderValue {
   // Header name.
-  string key = 1 [(validate.rules).string = {min_bytes: 1 max_bytes: 16384}];
+  string key = 1 [
+    (validate.rules).string = {min_bytes: 1 max_bytes: 16384 well_known_regex: HTTP_HEADER_NAME}
```
Comment (antoniovicente) [inline api/envoy/api/v2/core/base.proto]: Changes to this file may result in the rejection of configs that were previously considered valid.  Is there a way to disable validation for some subset of the fields?

Comment (htuch) [reply api/envoy/api/v2/core/base.proto]: I think fundamentally, @envoyproxy/api-shepherds need to decide if this is a breaking change. Arguably the RFC header name/value compliance was an implicit assumption already in the API and Envoy. Taking the opposite argument, there may be folks who relied on this previously unenforced behavior that will be affected. 

My $0.02 is that some aspects of this, e.g. eliminated \r\n\0, were already implied contract and this does not break for those. For other forbidden chars, I think this is a breaking change.

Comment (asraa) [reply api/envoy/api/v2/core/base.proto]: I think then that the right thing might be to replace the ones that will be checked for \r\n\0 with boilerplate custom regex for \r\n\0 with a deprecation and replacement field with the full validations... 
for the fields that did the full `headerIsValid` http utility checks, which match the full forbidden char set, those can probably stay

**Pull request merged**